### PR TITLE
Feat: Bulk actions

### LINKED ISSUE #3846: Add an option in the UI to select multiple conversations and change status
- [ ] https://github.com/chatwoot/chatwoot/issues/4525
- [ ] https://github.com/chatwoot/chatwoot/issues/4523
- [ ] https://github.com/chatwoot/chatwoot/issues/4524
- [ ] https://github.com/chatwoot/chatwoot/issues/4522

----

- [ ] Update the UI to select multiple conversations at once.
- [ ] Update the store / actions to perform bulk action.

### PR DESCRIPTION
**Pull request by fayazara**
# Pull Request Template

Demo
https://www.loom.com/share/6b55e165a17d498caae956fcd06bb617

## Description
This PR allows you to bulk update conversations.
1. Assigning conversations to agent
2. Resolving conversations

Fixes #3846 

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Test cases for Store, and QA.

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+        {{
+          $t('BULK_ACTION.CONVERSATIONS_SELECTED', {
+            conversationCount: conversations.length,
+          })
+        }}
+      </span>
+    </label>
+    <div class="bulk-action__actions flex-between">
+      <woot-button
+        v-tooltip="'Resolve'"
```
Comment (iamsivin) [inline app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue:21]: Use i18n here.

File: app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+        size="tiny"
+        variant="flat"
+        color-scheme="success"
+        icon="checkmark"
+        class="mr-small"
+        @click="resolveConversations"
+      />
+      <woot-button
+        v-tooltip="'Assign Agent'"
```
Comment (iamsivin) [inline app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue:30]: Use i18n here

File: app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+  height: 100%;
+  padding: var(--space-one);
+  p {
+    flex-grow: 1;
+  }
+  .agent-confirmation-actions {
+    width: 100%;
+    display: grid;
+    grid-template-columns: repeat(2, minmax(0, 1fr));
+    gap: 1rem;
```
Comment (iamsivin) [inline app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue:277]: Use css variables

File: app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+}
+
+.bulk-action__agents {
+  position: absolute;
+  bottom: 40px;
+  right: var(--space-small);
+  width: 100%;
+  box-shadow: 0 0.8rem 1.6rem rgb(50 50 93 / 8%),
+    0 0.4rem 1.2rem rgb(0 0 0 / 7%);
+  border-radius: 0.8rem;
```
Comment (iamsivin) [inline app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue:213]: Use CSS variables

File: app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+.mr-small {
+  margin-right: var(--space-smaller);
+}
+
+.bulk-action__agents {
+  position: absolute;
+  bottom: 40px;
+  right: var(--space-small);
+  width: 100%;
+  box-shadow: 0 0.8rem 1.6rem rgb(50 50 93 / 8%),
```
Comment (iamsivin) [inline app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue:211]: Can we use shadow as variable from here `app/javascript/shared/assets/stylesheets/shadows.scss`?

File: app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+.flex-between {
+  display: flex;
+  align-items: center;
+  justify-content: space-between;
+}
+.bulk-action__container {
+  padding: var(--space-normal) var(--space-one);
+  background-color: var(--s-50);
+  position: relative;
+  box-shadow: 6px 3px 22px 9px rgb(181 181 181 / 50%);
```
Comment (iamsivin) [inline app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue:180]: Can you reduce the color or increase the blur radius or remove the shadow from the right side? Looks darker.

image

**Review by iamsivin — COMMENTED** [review]
Please see the comments added.

**Comment by iamsivin** [comment]
Tested it locally LGTM👍🏻
https://www.loom.com/share/6735039a960f4de7be47797685b0c60c

File: app/javascript/dashboard/components/ChatList.vue [hunk 7]
```diff
@@ -96,6 +96,9 @@
         :chat="chat"
         :conversation-type="conversationType"
         :show-assignee="showAssigneeInConversationCard"
+        :selected="isConversationSelected(chat.id)"
+        @selectConversation="selectConversation"
```
Comment (nithindavid) [inline app/javascript/dashboard/components/ChatList.vue:100]:
```suggestion
        @select-conversation="selectConversation"
```

Comment (nithindavid) [reply app/javascript/dashboard/components/ChatList.vue:100]: Can we change all event names to snake case?

File: app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+      inboxes: 'inboxes/getInboxes',
+    }),
+    filteredAgents() {
+      return this.assignableAgents.filter(agent =>
+        agent.name.toLowerCase().includes(this.query.toLowerCase())
+      );
+    },
+    assignableAgents() {
+      return this.$store.getters['inboxAssignableAgents/getAssignableAgents'](
+        this.inboxes[0].id
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue:148]: Would this create a bug if there are no inboxes? Initially when there are no inboxes. Can you check?

File: app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+.flex-between {
+  display: flex;
+  align-items: center;
+  justify-content: space-between;
+}
+.bulk-action__container {
+  padding: var(--space-normal) var(--space-one);
+  background-color: var(--s-75);
+  position: relative;
+  box-shadow: 6px 3px 22px 9px rgb(181 181 181 / 25%);
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue:188]: Use shadows from `app/javascript/shared/assets/stylesheets/shadows.scss`

File: app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+    font-size: var(--font-size-mini);
+    margin-left: var(--space-smaller);
+  }
+  input[type='checkbox'] {
+    margin: var(--space-zero);
+    cursor: pointer;
+  }
+}
+
+.mr-small {
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue:210]: Move this to `app/javascript/dashboard/assets/scss/_utility-helpers.scss`

File: app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+  margin-right: var(--space-smaller);
+}
+
+.bulk-action__agents {
+  position: absolute;
+  bottom: 40px;
+  right: var(--space-small);
+  width: 100%;
+  box-shadow: 0 0.8rem 1.6rem rgb(50 50 93 / 8%),
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue:219]: Use shadow variable

File: app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+  flex-direction: column;
+  height: 100%;
+  padding: var(--space-one);
+  p {
+    flex-grow: 1;
+  }
+  .agent-confirmation-actions {
+    width: 100%;
+    display: grid;
+    grid-template-columns: repeat(2, minmax(0, 1fr));
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue:284]: 👍

File: app/javascript/dashboard/components/widgets/conversation/ConversationCard.vue [hunk 13]
```diff
@@ -5,11 +5,24 @@
...
+        :checked="selected"
+        class="checkbox"
+        type="checkbox"
+        @change="selectConversation($event.target.checked)"
+        @click.stop
+      />
+    </label>
     <thumbnail
-      v-if="!hideThumbnail"
+      v-if="!hideThumbnail && !hovered && !selected"
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/ConversationCard.vue:25]: Lets move this as a computed prop.

File: app/javascript/dashboard/components/widgets/conversation/ConversationCard.vue [hunk 14]
```diff
@@ -260,6 +281,18 @@ export default {
       }
       router.push({ path: frontendURL(path) });
     },
+    cardHover() {
+      if (!this.hideThumbnail) {
+        this.hovered = true;
+      } else this.hovered = false;
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/ConversationCard.vue:287]:
```suggestion
        this.hovered = !this.hideThumbnail;
```

File: app/javascript/dashboard/components/widgets/conversation/ConversationCard.vue [hunk 15]
```diff
@@ -260,6 +281,18 @@ export default {
       }
       router.push({ path: frontendURL(path) });
     },
+    cardHover() {
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/ConversationCard.vue:284]:
```suggestion
    onCardHover() {
```

File: app/javascript/dashboard/components/widgets/conversation/ConversationCard.vue [hunk 16]
```diff
@@ -260,6 +281,18 @@ export default {
...
+      if (!this.hideThumbnail) {
+        this.hovered = true;
+      } else this.hovered = false;
+    },
+    cardLeave() {
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/ConversationCard.vue:289]:
```suggestion
    onCardLeave() {
```

File: app/javascript/dashboard/components/widgets/conversation/ConversationCard.vue [hunk 17]
```diff
@@ -320,4 +357,22 @@ export default {
...
+  height: 40px;
+  width: 40px;
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  border-radius: 100%;
+  margin-top: var(--space-normal);
+  cursor: pointer;
+  &:hover {
+    background-color: rgb(194, 225, 255);
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/ConversationCard.vue:370]: Can we use a standard color?

**Review by nithindavid — CHANGES_REQUESTED** [review]
Please see the comments left

File: app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue (hunk 8 above)
Comment (fayazara) [reply app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue:148]: If there are no inboxes, there wouldn't be any conversations correct? @nithindavid

But, makes sense to handle this anyway.

File: app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue (hunk 9 above)
Comment (fayazara) [reply app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue:188]: I needed a shadow which was upward, hence used a custom one. I can add this in the css file and use it as a variable.

File: app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue (hunk 8 above)
Comment (pranavrajs) [reply app/javascript/dashboard/components/widgets/conversation/ConversationBulkActions.vue:148]: We will update this API, so not important.

**Comment by fayazara** [comment]
@nithindavid Updated the PR.

File: app/javascript/dashboard/store/modules/bulkActions.js [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+};
+
+export const getters = {
+  getUIFlags(_state) {
+    return _state.uiFlags;
+  },
+};
+
+export const actions = {
+  process: async function createAutomation({ commit }, payload) {
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/bulkActions.js:17]: createAutomation?

Comment (fayazara) [reply app/javascript/dashboard/store/modules/bulkActions.js:17]: Added by mistake, was supposed to be processAction.

Fixing

File: app/javascript/dashboard/i18n/locale/en/bulkActions.json [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+        "ASSIGN_CONFIRMATION_LABEL": "Are you sure you want to assign %{conversationCount} %{conversationLabel} to",
+        "GO_BACK_LABEL": "Go back",
+        "ASSIGN_LABEL": "Assign",
+        "ASSIGN_AGENT_TOOLTIP": "Assign Agent",
+        "RESOLVE_TOOLTIP": "Resolve",
+        "ASSIGN_SUCCESFUL": "Conversations assigned successfully",
+        "ASSIGN_FAILED": "Failed to assign conversations, please try again",
+        "RESOLVE_SUCCESFUL": "Conversations resolved successfully",
+        "RESOLVE_FAILED": "Failed to resolve conversations, please try again",
+        "ALL_CONVERSATIONS_SELECTED_ALERT": "Only conversations on this page are selected.",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/bulkActions.json:14]:
```suggestion
        "ALL_CONVERSATIONS_SELECTED_ALERT": "Conversations visible on this page are only selected.",
```

**Review by pranavrajs — APPROVED** [review]

**Pull request merged**